_posts/2020-07-05-gpg-card-with-google-vms.md

```diff
@@ -0,0 +1,110 @@
+    [SmartCard](https://en.wikipedia.org/wiki/OpenPGP_card) plus a reader.
+- The USB key/SmartCard configured (also) for SSH authentication (per links above).
+  - ```bash
+    $ gpg --card-status  # output should contain something similar to
+    Authentication key: 0EFC F3FD 3AFC 6D57 E0A4  1FB9 0C44 AEC0 3034 55E0
+    ```
+- [Configured gpg-agent to act as an SSH agent](https://www.bootc.net/archives/2013/06/09/my-perfect-gnupg-ssh-agent-setup/):
+  - ```bash
+    $ ssh-add -l  # should say something like
+    2048 SHA256:cDS0g8TiH0bJJthhJFW/GqKDmCnopkcEKSiRuEEAXmY cardno:000000012345 (RSA)
+    ```
+
+## TL;DR:
+
+```bash
+$ ssh-add -L > ~/.ssh/google_compute_engine.pub  # overwrite (!) the public key with OpenPGP one
+$ echo "fake content, actual private key on OpenPGP HW token" > ~/.ssh/google_compute_engine
+```
+
+## How Did We Get There
+
+`gcloud compute ssh` is just a tiny wrapper around `ssh` that does a couple of things on top of it:
+1. uses `~/.ssh/google_compute_engine(.pub)` SSH (public) key file path instead of standard
+   `~/.ssh/id_rsa(.pub)`,
+2. generates the key-pair automatically if it doesn't exist,
+3. updates target VM metadata (its `~/.ssh/authorized_keys`) with the public SSH key before
+   connecting (the convenient, and tricky, part).
+
+It should be possible to fool `gcloud compute ssh` by supplying the OpenPGP-backed public SSH key
+at the location it expects it, let's try:
+
+```bash
+$ ssh-add -L > ~/.ssh/google_compute_engine.pub
+$ gcloud compute ssh bencher
+WARNING: The private SSH key file for gcloud does not exist.
+WARNING: Your SSH key files are broken.
+private key (NOT FOUND) [/home/strohel/.ssh/google_compute_engine]
+public key  (OK)        [/home/strohel/.ssh/google_compute_engine.pub]
+We are going to overwrite all above files.
+Do you want to continue (y/N)?  N
+```
+
+Hmmm, `gcloud` was being a bit too smart here. We don't have the private key in file, because that's
+the entire point of confining it into a separate hardware device. Let's try to cheat a bit more:
+
+```bash
+$ touch ~/.ssh/google_compute_engine
+$ gcloud compute ssh bencher
+WARNING: The private SSH key file for gcloud is empty.
+(...)
+```
+
+Still no avail. Let's outsmart `gcloud` completely:
+
+```bash
+$ echo "fake content, actual private key on OpenPGP HW token" > ~/.ssh/google_compute_engine
+$ gcloud compute ssh bencher
+(...)
+Last login: ...
+```
+
+Yay. That's it, we need to write the public key in SSH format into `google_compute_engine.pub`, and
+some random content into `google_compute_engine`.
+
+## Notes
+
+1. I have tested this successfully with [YubiKey 4](https://support.yubico.com/support/solutions/articles/15000006486-yubikey-4)
+   and [ZeitControl "BasicCard" SmartCard (no longer) provided by FSFE](https://fsfe.org/news/2017/news-20171116-01.en.html).
+1. If you try to SSH without your OpenPGP token available, the error will look like:
```

Comment:
looks like `.1. is used for all the "bullet points" in `notes` section, probably err?

Reply:
checked the published blog, rendered properly so it's just me confused by some inner working of the framework

Reply:
Yeah, that's a feature of Markdown, see e.g. https://guides.github.com/features/mastering-markdown/ (Lists -> Ordered)

Quite handy when editing (with non-advanced editor).

Reply:
nice :)